driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java

```diff
@@ -29,357 +29,301 @@
-                    valueMap.put(valueEntry.getKey(), new MapCodec(codec, valueEntry.getValue(), protocolVersion));
-                primitiveMapsCodecs.put(type, valueMap);
-            }
-        }
-
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV2 = new PrimitiveCollectionCodecs(ProtocolVersion.V2);
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV3 = new PrimitiveCollectionCodecs(ProtocolVersion.V3);
-        private static final PrimitiveCollectionCodecs primitiveCollectionCodecsV4 = new PrimitiveCollectionCodecs(ProtocolVersion.V4);
-
-        static PrimitiveCollectionCodecs forVersion(ProtocolVersion version) {
-            // This happens during protocol negociation, when the version is not known yet.
-            // Use the smallest supported version, which is enough for what we need to do at this stage.
-            if (version == null)
-                version = ProtocolVersion.V1;
-
-            switch (version) {
-                case V1:
-                case V2:
-                    return primitiveCollectionCodecsV2;
-                case V3:
-                    return primitiveCollectionCodecsV3;
-                case V4:
-                    return primitiveCollectionCodecsV4;
-                default:
-                    throw version.unsupported();
-            }
-        }
+    protected TypeCodec(DataType cqlType, TypeToken<T> javaType) {
+        this.cqlType = cqlType;
+        this.javaType = javaType;
     }
 
+    /**
+     * Return the Java type that this codec deserializes to and serializes from.
+     * @return The Java type this codec deserializes to and serializes from.
+     */
+    public TypeToken<T> getJavaType() {
+        return javaType;
+    }
 
-    private TypeCodec() {}
-
-    public abstract T parse(String value);
-    public abstract String format(T value);
+    /**
+     * Return the CQL type that this codec deserializes from and serializes to.
+     * @return The Java type this codec deserializes from and serializes to.
+     */
+    public DataType getCqlType() {
+        return cqlType;
+    }
 
-    public abstract ByteBuffer serialize(T value);
-    public abstract T deserialize(ByteBuffer bytes);
+    /**
+     * Serialize the given value according to the CQL type
+     * handled by this codec.
+     * @param value An instance of T
+     * @param protocolVersion the protocol version to use when serializing
+     * {@code bytes}. In most cases, the proper value to provide for this argument
+     * is the value returned by {@link ProtocolOptions#getProtocolVersion} (which
+     * is the protocol version in use by the driver).
+     * @return A {@link ByteBuffer} instance containing the serialized form of T
+     * @throws InvalidTypeException if the given value does not have the expected type
+     * @throws NullPointerException if {@code value} is {@code null}, of for collections,
```

Comment:
Shouldn't it be legal for value to be null in some cases?  One use case I can think of is if you make a codec from `MappingCodec`, for example if you wanted to make a `TypeCodec` that mapped a `Type` to `Optional<Type>`.  This would allow you to map `null` results to `Optional.absent()` to make it more explicit that the user needs to handle those cases.

The problem is when you go to serialize an 'Optional.absent()' which would create a 'null' value that gets passed into CollectionCodec.serialize and throws an NPE.

Posted example codec and tests [here](https://gist.github.com/tolbertam/0bde5315cea099d4006e).   Both the `codec_named` and `codec_set` methods throw NPEs when passing in Optional.absent() while `codec_unset` throws an error `Invalid unset value for column c1` where I think it *should* work so maybe there is a bug there (but its entirely possible my understanding is wrong)